src/platform/Tizen/DnssdImpl.h

```diff
@@ -37,100 +42,103 @@ enum class ContextType
+    void * mCbContext;
+
+    dnssd_service_h mServiceHandle = 0;
+    bool mIsResolving              = false;
+
+    ResolveContext(DnssdTizen * instance, const char * name, const char * type, uint32_t interfaceId, DnssdResolveCallback callback,
+                   void * context);
+    ~ResolveContext() override;
 };
 
-class DnssdContexts
+class DnssdTizen
 {
 public:
-    DnssdContexts(const DnssdContexts &) = delete;
-    DnssdContexts & operator=(const DnssdContexts &) = delete;
-    ~DnssdContexts();
-    static DnssdContexts & GetInstance() { return sInstance; }
-
-    CHIP_ERROR Add(RegisterContext * context, dnssd_service_h service, const char * type, const char * name, uint16_t port,
-                   uint32_t interfaceId);
-    CHIP_ERROR Add(BrowseContext * context, dnssd_browser_h browser);
-    CHIP_ERROR Add(ResolveContext * context, dnssd_service_h service);
-    CHIP_ERROR Remove(GenericContext * context);
-    CHIP_ERROR Remove(const char * type, const char * name, uint16_t port, uint32_t interfaceId);
-    CHIP_ERROR Remove(ContextType type);
-    RegisterContext * Get(const char * type, const char * name, uint16_t port, uint32_t interfaceId);
+    DnssdTizen(const DnssdTizen &) = delete;
+    DnssdTizen & operator=(const DnssdTizen &) = delete;
 
-private:
-    DnssdContexts(){};
-    static DnssdContexts sInstance;
+    CHIP_ERROR Init(DnssdAsyncReturnCallback initCallback, DnssdAsyncReturnCallback errorCallback, void * context);
+    CHIP_ERROR Shutdown();
+
+    CHIP_ERROR RegisterService(const DnssdService & service, DnssdPublishCallback callback, void * context);
+    CHIP_ERROR UnregisterAllServices();
+
+    CHIP_ERROR Browse(const char * type, DnssdServiceProtocol protocol, chip::Inet::IPAddressType addressType,
+                      chip::Inet::InterfaceId interface, DnssdBrowseCallback callback, void * context);
 
-    void Delete(GenericContext * context);
-    std::vector<GenericContext *> mContexts;
+    CHIP_ERROR Resolve(const DnssdService & browseResult, chip::Inet::InterfaceId interface, DnssdResolveCallback callback,
+                       void * context);
+
+    // TODO (a.bokowy): Make this method private
+    CHIP_ERROR RemoveContext(GenericContext * context);
+
+    static DnssdTizen & GetInstance() { return sInstance; }
+
+private:
+    DnssdTizen() = default;
+    static DnssdTizen sInstance;
+
+    RegisterContext * CreateRegisterContext(const char * type, const DnssdService & service, DnssdPublishCallback callback,
+                                            void * context);
+    BrowseContext * CreateBrowseContext(const char * type, DnssdServiceProtocol protocol, uint32_t interfaceId,
+                                        DnssdBrowseCallback callback, void * context);
+    ResolveContext * CreateResolveContext(const char * name, const char * type, uint32_t interfaceId, DnssdResolveCallback callback,
+                                          void * context);
+
+    std::mutex mMutex;
```

Comment:
I didn't see thread considerations mentioned in the pull request description.  Can you describe how the locking works, why it's needed, and what was previously broken with it missing?

Reply:
Every Tizen native API callback is run in a dedicated thread. The flow is as follows:

1. Call to `ChipDnssdPublishService()`
2. Call to `DnssdTizen::GetInstance().RegisterService()`
3. The `RegisterService()` function prepares Tizen Native API `dnssd_service_h` for registration
4. The `RegisterService()` function calls `MainLoop::Instance().AsyncRequest(RegisterAsync, serviceCtx)`
5. The `AsyncRequest()` creates new GLib main loop context and runs `RegisterAsync()` (at this point we are still in the "main" thread)
6. The `RegisterAsync()` function registers new mDNS service using `dnssd_register_local_service()`, where the `dnssd_register_local_service` adds some callbacks to the current default GLib context (all I/O will be performed in that context)
7. In case of `RegisterAsync()` success the `AsyncRequest()` creates new thread and runs GLlib main loop dispatcher in it, with the context previously created
8. In created thread the mDNS I/O is processed and in case of successful service registration (e.g. checks for conflicts) the callback registered with `dnssd_register_local_service` is called - the `OnRegister` function
9. The `OnRegister` (run in a dedicated thread) might remove context from the `mContexts` using `rCtx->mInstance->RemoveContext()`

Similar scenario is for browse and resolve. So, the `mContexts` is accessed from multiple threads. Hence, it is required to protect it with the mutex.